cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +682,81 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-// s.mu must be locked when calling handleProcessExit
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - for a given container, the init process exit MUST be the last exit published
+// This is achieved by:
+// - waiting for all pending execs to either successfully start or early-exit
+// (this happens before handleInitExit is called)
+// - killing all running container processes
+// - removing all processes from s.running so that we s.processExits doesn't
+// publish their exits
+// - subscribing to exits+waiting for the running process count (for a
+// container) to go to 0, or timing out.
+// - publishing the received exec exits, and then publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
+		log.G(s.context).WithError(err).WithField("id", p.ID()).
+			Error("failed to kill init's children")
+	}
+
+	go func() {
+		timeout := 20 * time.Millisecond
+		timer := time.NewTimer(timeout)
+		var numRunningExecs int
+		defer timer.Stop()
+		// wait for running processes to exit
+		for {
+			<-timer.C
+
+			s.lifecycleMu.Lock()
+			newNumRunningExecs := s.runningExecs[c]
+			s.lifecycleMu.Unlock()
+
+			if newNumRunningExecs == 0 {
+				break
+			}
+
+			if newNumRunningExecs == numRunningExecs {
+				break
 			}
+
+			numRunningExecs = newNumRunningExecs
+
+			// timer channel has been drained, so it's safe
+			// to just call Reset
```

Comment:
You'd think so, but it is not. The timer is still running in the background, so another timeout might have elapsed again if the goroutine gets preempted between the channel receive and here. You always have to stop, drain, then reset.

Reply:
Ohhhhh, 🤦‍♀️ you linked to the issues before, but I thought since we were draining it (once) we'd be fine. I understand now 🙃